lib/python/qmk/cli/new/keyboard.py

```diff
@@ -0,0 +1,153 @@
+"""This script automates creating the files necessary for a new QMK keyboard.
+"""
+import shutil
+import re
+
+from milc import cli
+from pathlib import Path
+from datetime import datetime
+
+from qmk.constants import SUPPORTED_MCUS
+from qmk.keyboard import generate_pid
+from qmk.questions import question
+from qmk.keymap import rewrite_source
+
+
+@cli.argument('-kb', '--keyboard', help='Specify keyboard name. Example: clueboard/66')
+@cli.argument('-pr', '--project', help='The human-friendly name of the keyboard. Example: Clueboard 66%%')
+@cli.argument('-mcu', '--microcontroller', help='Specify the microcontroller used for the keyboard.')
+@cli.argument('-u', '--username', help='Your GitHub username. Will be pasted into generated files.')
+@cli.subcommand('Creates a new keyboard project.')
+def new_keyboard(cli):
+    """Creates a new keyboard project.
+    """
+
+    # Root path for template files
+    template_root_path = Path("quantum/template")
+    template_base_path = template_root_path / "base"
+    year = datetime.now().year
+    valid_keyboard_name = re.compile(r'^[a-z0-9][a-z0-9_/]+$')
+
+    if cli.args.keyboard:
+        keyboard = keyboard_name = cli.args.keyboard
+    else:
+        # ask for user input if keyboard was not provided in the command line
+        cli.echo("""What is the keyboard's project name?
+
+    This will be the name used to compile firmware for your keyboard. The
+    project name may only contain lower case letters, numbers, and the
+    underscore character, and must not begin with an underscore.
+
+    Files will be placed in `qmk_firmware/keyboards/<project_name>/`.""")
+        keyboard = question("Project Name: ")
+        #keyboard = Path(re.sub(r'[^a-z0-9_/]', "", keyboard.lower()))
+        keyboard = Path(keyboard.lower())
+
+
+    if cli.args.project:
+        keyboard_name = cli.args.project
+    else:
+        # Ask the common name of the keyboard, if not specified by --project
+        cli.echo("""What is the keyboard's common name?
+
+    This is the name that people will use to refer to your keyboard. It should
+    be something human-friendly, like \"Clueboard 66%%\" or \"Ergodox EZ\".""")
+        keyboard_name = question("Common Name: ")
+
+
+    if cli.args.microcontroller:
+        mcu = cli.args.microcontroller
+        # Create a case-insensitive regular expression for checking the -mcu
+        #   argument
+        mcu_pattern = re.compile(mcu, re.IGNORECASE)
+
+        # Loop through the list of supported MCUs
+        for i in range(len(SUPPORTED_MCUS)):
+            # If the value of -mcu is a case-insensitive match for a key in
+            #   SUPPORTED_MCUS...
+            if mcu_pattern.fullmatch(list(SUPPORTED_MCUS.keys())[i]):
+                # ... set the mcu to its proper name...
+                mcu = list(SUPPORTED_MCUS.keys())[i]
+                # ... and assign its architecture
+                arch = SUPPORTED_MCUS[mcu]
+                break
```

Comment:
I might miss something, but I think you are overcomplicating things. :D
Try:
```
mcu = cli.args.microcontroller.casefold()
if mcu in SUPPORTED_MCUS:
    arch = SUPPORTED_MCUS[mcu]
```

Reply:
It seems like it *should* work, but it doesn't match anything when I use it. :man_shrugging:

Reply:
Yeah, you are right.
Btw, is there a specific reason to mix cases in `SUPPORTED_MCUS`?

Reply:
This is for the `-mcu` argument. Because the MCU list isn't consistent about case (AVRs being lowercase and STM32s being uppercase), I thought it more user-friendly to make the argument case-insensitive, then it doesn't matter how the MCU is given – `-mcu stm32f303` works as expected.

Reply:
I get that. What I don't get is why the MCU list is inconsistent. If every entry was either lower or upper case, this user-friendly case-insensitive input would be much simpler. 

So I don't see the benefit of mixing cases in the MCU list, but I might miss some piece of the puzzle.

Reply:
There's no benefit, it's just something we have to live with until someone (probably me) makes the STM32 MCUs lowercase as a breaking change: https://github.com/qmk/qmk_firmware/blob/master/quantum/mcu_selection.mk

Reply:
Got it. 

In this case, I'd recommend this: 

``` 
# AVRs are lowercase 
if mcu.lower() in SUPPORTED_MCUS: 
    mcu = mcu.lower()
    arch = SUPPORTED_MCUS[mcu]

# STMs are lowercase 
if mcu.upper() in SUPPORTED_MCUS: 
    mcu = mcu.upper()
    arch = SUPPORTED_MCUS[mcu]
``` 

We know the arch from the case at the moment, but I'd still get it from the dictionary, so if STMs are changed to lowercase in the future, it will not break this code. 

(note: the formatting might be off, I'm from phone)

Reply:
6eb6210